pkg/util/writerlease/writerlease.go

```diff
@@ -0,0 +1,295 @@
+	defer l.lock.Unlock()
+	l.expires = time.Time{}
+}
+
+func (l *WriterLease) Wait() bool {
+	<-l.once
+	state, _, _ := l.leaseState()
+	return state == Leader
+}
+
+func (l *WriterLease) WaitUntil(t time.Duration) (bool, bool) {
+	select {
+	case <-l.once:
+	case <-time.After(t):
+		return false, false
+	}
+	state, _, _ := l.leaseState()
+	return state == Leader, true
+}
+
+func (l *WriterLease) Try(key string, fn WorkFunc) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	l.queued[key] = fn
+	if l.state == Follower {
+		l.queue.AddAfter(key, l.expires.Sub(nowFn()))
+	} else {
+		l.queue.Add(key)
+	}
+}
+
+func (l *WriterLease) Extend(key string) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	if _, ok := l.queued[key]; ok {
+		delete(l.queued, key)
+		switch l.state {
+		case Follower:
+			l.tick++
+			backoff := l.nextBackoff()
+			glog.V(4).Infof("[%s] Clearing work for %s and extending lease by %s", l.name, key, backoff)
+			l.expires = nowFn().Add(backoff)
+		}
+	}
+}
+
+func (l *WriterLease) Len() int {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	return len(l.queued)
+}
+
+func (l *WriterLease) Remove(key string) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	delete(l.queued, key)
+}
+
+func (l *WriterLease) get(key string) WorkFunc {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	return l.queued[key]
+}
+
+func (l *WriterLease) leaseState() (State, time.Time, int) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	return l.state, l.expires, l.tick
+}
+
+func (l *WriterLease) work() {
+	for {
+		item, shutdown := l.queue.Get()
+		if shutdown {
+			return
+		}
+		key := item.(string)
+
+		fn := l.get(key)
+		if fn == nil {
+			glog.V(4).Infof("[%s] Work item %s was cleared, done", l.name, key)
+			l.queue.Done(key)
+			continue
+		}
+
+		leaseState, leaseExpires, _ := l.leaseState()
+		if leaseState == Follower {
+			// if we are following, continue to defer work until the lease expires
+			if remaining := leaseExpires.Sub(nowFn()); remaining > 0 {
+				glog.V(4).Infof("[%s] Follower, %s remaining in lease", l.name, remaining)
+				l.queue.AddAfter(key, remaining)
+				l.queue.Done(key)
+				continue
+			}
+			glog.V(4).Infof("[%s] Lease expired, running %s", l.name, key)
+		} else {
+			glog.V(4).Infof("[%s] Lease owner or electing, running %s", l.name, key)
+		}
+
+		isLeader, retry := fn()
+		if retry {
+			// come back in a bit
+			glog.V(4).Infof("[%s] Retrying %s", l.name, key)
+			l.queue.AddAfter(key, l.retryInterval)
+			l.queue.Done(key)
+			continue
+		}
+
+		l.finishKey(key, isLeader)
+	}
+}
+
+func (l *WriterLease) finishKey(key string, isLeader bool) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	resolvedElection := l.state == Election
+	if isLeader {
+		switch l.state {
+		case Election, Follower:
+			l.tick = 0
+			l.state = Leader
+		}
+		l.expires = nowFn().Add(l.maxBackoff)
+	} else {
+		switch l.state {
+		case Election, Leader:
+			l.tick = 0
+			l.state = Follower
+		case Follower:
+			l.tick++
+		}
+		l.expires = nowFn().Add(l.nextBackoff())
+	}
```

Comment:
I am going to have to look at this again ... where is the actual election done here? Are we setting this (aka doing the election) based on the status update for the route going through? That level of indirection might make it tough to understand this without some comments here.

Reply:
If you are able to write to the object (because you have a delta to the current state) you take ownership of the lease.  If you get a contention, or if you observe another client write to the object, you go into random exponential backoff until one of your writes succeeds (resetting you to zero backoff) or you hit the max backoff (the lease interval).

If you're in follower mode, everytime you observe another client doing real work you extend the lease.

For three router processes, the first route to be admitted would create a status attempt from each router (all in election).  Which ever won would consider itself the leader - the others would go into backoff.  If the lease interval expires without the followers observing any work getting done, they'll start trying to do work.  If an entire backoff interval goes by and there is no work, the lease is released and they'll start competing again.  1m is completely unscientific but is roughly equivalent to our current mechanism.